Move xla codegen to aten.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45241 Move xla codegen to aten.**

Differential Revision: [D23926750](https://our.internmc.facebook.com/intern/diff/D23926750)

Commit:
Move xla codegen to aten.

[ghstack-poisoned]

ailzhang commented:
Thanks to @ljk53 's https://github.com/pytorch/pytorch/pull/45137, we have enough information in aten for xla backend extension! 
ps: #45137 is only the first PR to bring dispatchStub information in native_functions.yaml, but XLA only overrides a subset of our ops so fortunately it's fully covered by the first PR. 
Btw I haven't removed the codegen in `gen_variable_type.py` yet since there might be other users using it atm. My plan is to move xla immediately to this new file and give some time for other users to migrate before I remove it.

ezyang commented:
Don't forget to check in the template? (Unless it's already there lol)

ezyang commented on tools/codegen/gen.py:
```diff
@@ -814,6 +814,17 @@ def compute_declaration_yaml(f: NativeFunction) -> object:
 
+@with_native_function
+def compute_registration_declarations(f: NativeFunction) -> str:
+    name = cpp.name(f.func)
```
Not really the fault of this PR, but maybe we should add `dispatcher.name` and point it at `cpp.name`; would help make this look more uniform

Commit:
Update on "Move xla codegen to aten."




[ghstack-poisoned]

ezyang commented on aten/src/ATen/templates/RegistrationDeclarations.h:
```diff
@@ -0,0 +1,4 @@
+// This file contains all native_functions that can be registered to
+// and the schema string that they should be registered with
```
Interesting that you don't need any includes here

ailzhang replied:
Aha because we don't include this file directly... There're some additional stuff xla want to set up in the header file about class/namespace so we generate a new file to include. But yea it's possible to make this a real header file for backend extensions.

Commit:
Update on "Move xla codegen to aten."


Differential Revision: [D23926750](https://our.internmc.facebook.com/intern/diff/D23926750)

[ghstack-poisoned]
